File: old/tags/2.2/java/edu/nps/moves/dis/GriddedDataPdu.java
package edu.nps.moves.dis;

import java.util.*;
import java.io.*;
import edu.nps.moves.jaxb.dis.*;

/**
 * Section 5.3.11.2: Information about globat, spatially varying enviornmental effects. This requires manual cleanup; the grid axis        records are variable sized. UNFINISHED
 *
 * Copyright (c) 2007, MOVES Institute, Naval Postgraduate School. All rights reserved.
 * This work is licensed under the BSD open source license, available at https://www.movesinstitute.org/licenses/bsd.html
 *
 * @author DMcG
 */
public class GriddedDataPdu extends SyntheticEnvironmentFamilyPdu implements Serializable
{
   /** environmental simulation application ID */
   protected EntityID  environmentalSimulationApplicationID = new EntityID(); 

   /** unique identifier for each piece of enviornmental data */
   protected int  fieldNumber;

   /** sequence number for the total set of PDUS used to transmit the data */
   protected int  pduNumber;

   /** Total number of PDUS used to transmit the data */
   protected int  pduTotal;

   /** coordinate system of the grid */
   protected int  coordinateSystem;

   /** number of grid axes for the environmental data */
   protected short  numberOfGridAxes;

   /** are domain grid axes identidal to those of the priveious domain update? */
   protected short  constantGrid;

   /** type of environment */
   protected EntityType  environmentType = new EntityType(); 

   /** orientation of the data grid */
   protected Orientation  orientation = new Orientation(); 

   /** valid time of the enviormental data sample, 64 bit unsigned int */
   protected long  sampleTime;

   /** total number of all data values for all pdus for an environmental sample */
   protected long  totalValues;

   /** total number of data values at each grid point. */
   protected short  vectorDimension;

   /** padding */
   protected int  padding1;

   /** padding */
   protected short  padding2;

   /** Grid data @@@This is wrong */
   protected List gridDataList = new ArrayList(); 

/** Constructor */
 public GriddedDataPdu()
 {
    setPduType( (short)42 );
 }

/** 
 * Constructor--takes a parallel jaxb object and returns an open-dis object 
 * 1.4_sed_bait_start */
 public GriddedDataPdu(edu.nps.moves.jaxb.dis.GriddedDataPdu x)
 {
     super(x); // Call superclass constructor


     edu.nps.moves.dis.EntityID foo_0;
     if(x.getEnvironmentalSimulationApplicationID() == null)
        foo_0 = new edu.nps.moves.dis.EntityID();
      else
        foo_0 = new edu.nps.moves.dis.EntityID(x.getEnvironmentalSimulationApplicationID() );
     this.setEnvironmentalSimulationApplicationID(foo_0);

     this.fieldNumber = x.getFieldNumber();
     this.pduNumber = x.getPduNumber();
     this.pduTotal = x.getPduTotal();
     this.coordinateSystem = x.getCoordinateSystem();
     this.numberOfGridAxes = x.getNumberOfGridAxes();
     this.constantGrid = x.getConstantGrid();

     edu.nps.moves.dis.EntityType foo_7;
     if(x.getEnvironmentType() == null)
        foo_7 = new edu.nps.moves.dis.EntityType();
      else
        foo_7 = new edu.nps.moves.dis.EntityType(x.getEnvironmentType() );
     this.setEnvironmentType(foo_7);


     edu.nps.moves.dis.Orientation foo_8;
     if(x.getOrientation() == null)
        foo_8 = new edu.nps.moves.dis.Orientation();
      else
        foo_8 = new edu.nps.moves.dis.Orientation(x.getOrientation() );
     this.setOrientation(foo_8);

     this.sampleTime = x.getSampleTime();
     this.totalValues = x.getTotalValues();
     this.vectorDimension = x.getVectorDimension();
     this.padding1 = x.getPadding1();
     this.padding2 = x.getPadding2();
     this.gridDataList = new ArrayList();
     for(int idx = 0; idx < x.getGridDataList().size(); idx++)
     {
        this.gridDataList.add( new edu.nps.moves.dis.GridAxisRecord((edu.nps.moves.jaxb.dis.GridAxisRecord) x.getGridDataList().get(idx)));
     }
 }
/* 1.4_sed_bait_end */


/**
 * returns a jaxb object intialized from this object, given an empty jaxb object
 * 1.4_sed_bait_start **/
 public edu.nps.moves.jaxb.dis.GriddedDataPdu initializeJaxbObject(edu.nps.moves.jaxb.dis.GriddedDataPdu x)
 {
     super.initializeJaxbObject(x); // Call superclass initializer

     ObjectFactory factory = new ObjectFactory();

     x.setEnvironmentalSimulationApplicationID( this.getEnvironmentalSimulationApplicationID().initializeJaxbObject(factory.createEntityID()) );
     x.setFieldNumber( this.getFieldNumber() );
     x.setPduNumber( this.getPduNumber() );
     x.setPduTotal( this.getPduTotal() );
     x.setCoordinateSystem( this.getCoordinateSystem() );
     x.setNumberOfGridAxes( this.getNumberOfGridAxes() );
     x.setConstantGrid( this.getConstantGrid() );
     x.setEnvironmentType( this.getEnvironmentType().initializeJaxbObject(factory.createEntityType()) );
     x.setOrientation( this.getOrientation().initializeJaxbObject(factory.createOrientation()) );
     x.setSampleTime( this.getSampleTime() );
     x.setTotalValues( this.getTotalValues() );
     x.setVectorDimension( this.getVectorDimension() );
     x.setPadding1( this.getPadding1() );
     x.setPadding2( this.getPadding2() );

     List gridDataList_1 = x.getGridDataList();
     for(int idx = 0; idx < gridDataList.size(); idx++)
     {
         GridAxisRecord a = (edu.nps.moves.dis.GridAxisRecord)gridDataList.get(idx);
         gridDataList_1.add(a.initializeJaxbObject(factory.createGridAxisRecord()));
     }
   return x;
 }
/* 1.4_sed_bait_end */


public int getMarshalledSize()
{
   int marshalSize = 0; 

   marshalSize = super.getMarshalledSize();
   marshalSize = marshalSize + environmentalSimulationApplicationID.getMarshalledSize();  // environmentalSimulationApplicationID
   marshalSize = marshalSize + 2;  // fieldNumber
   marshalSize = marshalSize + 2;  // pduNumber
   marshalSize = marshalSize + 2;  // pduTotal
   marshalSize = marshalSize + 2;  // coordinateSystem
   marshalSize = marshalSize + 1;  // numberOfGridAxes
   marshalSize = marshalSize + 1;  // constantGrid
   marshalSize = marshalSize + environmentType.getMarshalledSize();  // environmentType
   marshalSize = marshalSize + orientation.getMarshalledSize();  // orientation
   marshalSize = marshalSize + 8;  // sampleTime
   marshalSize = marshalSize + 4;  // totalValues
   marshalSize = marshalSize + 1;  // vectorDimension
   marshalSize = marshalSize + 2;  // padding1
   marshalSize = marshalSize + 1;  // padding2
   for(int idx=0; idx < gridDataList.size(); idx++)
   {
        GridAxisRecord listElement = (GridAxisRecord)gridDataList.get(idx);
        marshalSize = marshalSize + listElement.getMarshalledSize();
   }

   return marshalSize;
}


public void setEnvironmentalSimulationApplicationID(EntityID pEnvironmentalSimulationApplicationID)
{ environmentalSimulationApplicationID = pEnvironmentalSimulationApplicationID;
}

public EntityID getEnvironmentalSimulationApplicationID()
{ return environmentalSimulationApplicationID; }

public void setFieldNumber(int pFieldNumber)
{ fieldNumber = pFieldNumber;
}

public int getFieldNumber()
{ return fieldNumber; 
}

public void setPduNumber(int pPduNumber)
{ pduNumber = pPduNumber;
}

public int getPduNumber()
{ return pduNumber; 
}

public void setPduTotal(int pPduTotal)
{ pduTotal = pPduTotal;
}

public int getPduTotal()
{ return pduTotal; 
}

public void setCoordinateSystem(int pCoordinateSystem)
{ coordinateSystem = pCoordinateSystem;
}

public int getCoordinateSystem()
{ return coordinateSystem; 
}

public short getNumberOfGridAxes()
{ return (short)gridDataList.size();
}

/** Note that setting this value will ot change the marshalled value. The list whose length this describes is used for that purpose.
 * The getnumberOfGridAxes method will also be based on the actual list length rather than this value. 
 * The method is simply here for java bean completeness.
 */
public void setNumberOfGridAxes(short pNumberOfGridAxes)
{ numberOfGridAxes = pNumberOfGridAxes;
}

public void setConstantGrid(short pConstantGrid)
{ constantGrid = pConstantGrid;
}

public short getConstantGrid()
{ return constantGrid; 
}

public void setEnvironmentType(EntityType pEnvironmentType)
{ environmentType = pEnvironmentType;
}

public EntityType getEnvironmentType()
{ return environmentType; }

public void setOrientation(Orientation pOrientation)
{ orientation = pOrientation;
}

public Orientation getOrientation()
{ return orientation; }

public void setSampleTime(long pSampleTime)
{ sampleTime = pSampleTime;
}

public long getSampleTime()
{ return sampleTime; 
}

public void setTotalValues(long pTotalValues)
{ totalValues = pTotalValues;
}

public long getTotalValues()
{ return totalValues; 
}

public void setVectorDimension(short pVectorDimension)
{ vectorDimension = pVectorDimension;
}

public short getVectorDimension()
{ return vectorDimension; 
}

public void setPadding1(int pPadding1)
{ padding1 = pPadding1;
}

public int getPadding1()
{ return padding1; 
}

public void setPadding2(short pPadding2)
{ padding2 = pPadding2;
}

public short getPadding2()
{ return padding2; 
}

public void setGridDataList(List pGridDataList)
{ gridDataList = pGridDataList;
}

public List getGridDataList()
{ return gridDataList; }


public void marshal(DataOutputStream dos)
{
    super.marshal(dos);
    try 
    {
       environmentalSimulationApplicationID.marshal(dos);
       dos.writeShort( (short)fieldNumber);
       dos.writeShort( (short)pduNumber);
       dos.writeShort( (short)pduTotal);
       dos.writeShort( (short)coordinateSystem);
       dos.writeByte( (byte)gridDataList.size());
       dos.writeByte( (byte)constantGrid);
       environmentType.marshal(dos);
       orientation.marshal(dos);
       dos.writeLong( (long)sampleTime);
       dos.writeInt( (int)totalValues);
       dos.writeByte( (byte)vectorDimension);
       dos.writeShort( (short)padding1);
       dos.writeByte( (byte)padding2);

       for(int idx = 0; idx < gridDataList.size(); idx++)
       {
            GridAxisRecord aGridAxisRecord = (GridAxisRecord)gridDataList.get(idx);
            aGridAxisRecord.marshal(dos);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      System.out.println(e);}
    } // end of marshal method

public void unmarshal(DataInputStream dis)
{
    super.unmarshal(dis);

    try 
    {
       environmentalSimulationApplicationID.unmarshal(dis);
       fieldNumber = dis.readShort();
       pduNumber = dis.readShort();
       pduTotal = dis.readShort();
       coordinateSystem = dis.readShort();
       numberOfGridAxes = dis.readByte();
       constantGrid = dis.readByte();
       environmentType.unmarshal(dis);
       orientation.unmarshal(dis);
       sampleTime = dis.readLong();
       totalValues = dis.readInt();
       vectorDimension = dis.readByte();
       padding1 = dis.readShort();
       padding2 = dis.readByte();
        for(int idx = 0; idx < numberOfGridAxes; idx++)
        {
           GridAxisRecord anX = new GridAxisRecord();
            anX.unmarshal(dis);
            gridDataList.add(anX);
        };

    } // end try 
   catch(Exception e)
    { 
      System.out.println(e); 
    }
 } // end of unmarshal method 


 /**
  * The equals method doesn't always work--mostly on on classes that consist only of primitives. Be careful.
  */
 public boolean equals(GriddedDataPdu rhs)
 {
     boolean ivarsEqual = true;

    if(rhs.getClass() != this.getClass())
        return false;

     if( ! (environmentalSimulationApplicationID.equals( rhs.environmentalSimulationApplicationID) )) ivarsEqual = false;
     if( ! (fieldNumber == rhs.fieldNumber)) ivarsEqual = false;
     if( ! (pduNumber == rhs.pduNumber)) ivarsEqual = false;
     if( ! (pduTotal == rhs.pduTotal)) ivarsEqual = false;
     if( ! (coordinateSystem == rhs.coordinateSystem)) ivarsEqual = false;
     if( ! (numberOfGridAxes == rhs.numberOfGridAxes)) ivarsEqual = false;
     if( ! (constantGrid == rhs.constantGrid)) ivarsEqual = false;
     if( ! (environmentType.equals( rhs.environmentType) )) ivarsEqual = false;
     if( ! (orientation.equals( rhs.orientation) )) ivarsEqual = false;
     if( ! (sampleTime == rhs.sampleTime)) ivarsEqual = false;
     if( ! (totalValues == rhs.totalValues)) ivarsEqual = false;
     if( ! (vectorDimension == rhs.vectorDimension)) ivarsEqual = false;
     if( ! (padding1 == rhs.padding1)) ivarsEqual = false;
     if( ! (padding2 == rhs.padding2)) ivarsEqual = false;

     for(int idx = 0; idx < gridDataList.size(); idx++)
     {
        GridAxisRecord x = (GridAxisRecord)gridDataList.get(idx);
        if( ! ( gridDataList.get(idx).equals(rhs.gridDataList.get(idx)))) ivarsEqual = false;
     }


    return ivarsEqual;
 }
} // end of class

File: old/tags/2.1/cpp/DIS/GriddedDataPdu.cpp
#include <DIS/GriddedDataPdu.h> 

using namespace DIS;


GriddedDataPdu::GriddedDataPdu() : SyntheticEnvironmentFamilyPdu(),
   _environmentalSimulationApplicationID(), 
   _fieldNumber(0), 
   _pduNumber(0), 
   _pduTotal(0), 
   _coordinateSystem(0), 
   _numberOfGridAxes(0), 
   _constantGrid(0), 
   _environmentType(), 
   _orientation(), 
   _sampleTime(0), 
   _totalValues(0), 
   _vectorDimension(0), 
   _padding1(0), 
   _padding2(0)
{
    setPduType( 42 );
}

GriddedDataPdu::~GriddedDataPdu()
{
    _gridDataList.clear();
}

EntityID& GriddedDataPdu::getEnvironmentalSimulationApplicationID() 
{
    return _environmentalSimulationApplicationID;
}

const EntityID& GriddedDataPdu::getEnvironmentalSimulationApplicationID() const
{
    return _environmentalSimulationApplicationID;
}

void GriddedDataPdu::setEnvironmentalSimulationApplicationID(const EntityID &pX)
{
    _environmentalSimulationApplicationID = pX;
}

unsigned short GriddedDataPdu::getFieldNumber() const
{
    return _fieldNumber;
}

void GriddedDataPdu::setFieldNumber(unsigned short pX)
{
    _fieldNumber = pX;
}

unsigned short GriddedDataPdu::getPduNumber() const
{
    return _pduNumber;
}

void GriddedDataPdu::setPduNumber(unsigned short pX)
{
    _pduNumber = pX;
}

unsigned short GriddedDataPdu::getPduTotal() const
{
    return _pduTotal;
}

void GriddedDataPdu::setPduTotal(unsigned short pX)
{
    _pduTotal = pX;
}

unsigned short GriddedDataPdu::getCoordinateSystem() const
{
    return _coordinateSystem;
}

void GriddedDataPdu::setCoordinateSystem(unsigned short pX)
{
    _coordinateSystem = pX;
}

unsigned char GriddedDataPdu::getNumberOfGridAxes() const
{
   return _gridDataList.size();
}

unsigned char GriddedDataPdu::getConstantGrid() const
{
    return _constantGrid;
}

void GriddedDataPdu::setConstantGrid(unsigned char pX)
{
    _constantGrid = pX;
}

EntityType& GriddedDataPdu::getEnvironmentType() 
{
    return _environmentType;
}

const EntityType& GriddedDataPdu::getEnvironmentType() const
{
    return _environmentType;
}

void GriddedDataPdu::setEnvironmentType(const EntityType &pX)
{
    _environmentType = pX;
}

Orientation& GriddedDataPdu::getOrientation() 
{
    return _orientation;
}

const Orientation& GriddedDataPdu::getOrientation() const
{
    return _orientation;
}

void GriddedDataPdu::setOrientation(const Orientation &pX)
{
    _orientation = pX;
}

long GriddedDataPdu::getSampleTime() const
{
    return _sampleTime;
}

void GriddedDataPdu::setSampleTime(long pX)
{
    _sampleTime = pX;
}

unsigned int GriddedDataPdu::getTotalValues() const
{
    return _totalValues;
}

void GriddedDataPdu::setTotalValues(unsigned int pX)
{
    _totalValues = pX;
}

unsigned char GriddedDataPdu::getVectorDimension() const
{
    return _vectorDimension;
}

void GriddedDataPdu::setVectorDimension(unsigned char pX)
{
    _vectorDimension = pX;
}

unsigned short GriddedDataPdu::getPadding1() const
{
    return _padding1;
}

void GriddedDataPdu::setPadding1(unsigned short pX)
{
    _padding1 = pX;
}

unsigned char GriddedDataPdu::getPadding2() const
{
    return _padding2;
}

void GriddedDataPdu::setPadding2(unsigned char pX)
{
    _padding2 = pX;
}

std::vector<GridAxisRecord>& GriddedDataPdu::getGridDataList() 
{
    return _gridDataList;
}

const std::vector<GridAxisRecord>& GriddedDataPdu::getGridDataList() const
{
    return _gridDataList;
}

void GriddedDataPdu::setGridDataList(const std::vector<GridAxisRecord>& pX)
{
     _gridDataList = pX;
}

void GriddedDataPdu::marshal(DataStream& dataStream) const
{
    SyntheticEnvironmentFamilyPdu::marshal(dataStream); // Marshal information in superclass first
    _environmentalSimulationApplicationID.marshal(dataStream);
    dataStream << _fieldNumber;
    dataStream << _pduNumber;
    dataStream << _pduTotal;
    dataStream << _coordinateSystem;
    dataStream << ( unsigned char )_gridDataList.size();
    dataStream << _constantGrid;
    _environmentType.marshal(dataStream);
    _orientation.marshal(dataStream);
    dataStream << _sampleTime;
    dataStream << _totalValues;
    dataStream << _vectorDimension;
    dataStream << _padding1;
    dataStream << _padding2;

     for(size_t idx = 0; idx < _gridDataList.size(); idx++)
     {
        GridAxisRecord x = _gridDataList[idx];
        x.marshal(dataStream);
     }

}

void GriddedDataPdu::unmarshal(DataStream& dataStream)
{
    SyntheticEnvironmentFamilyPdu::unmarshal(dataStream); // unmarshal information in superclass first
    _environmentalSimulationApplicationID.unmarshal(dataStream);
    dataStream >> _fieldNumber;
    dataStream >> _pduNumber;
    dataStream >> _pduTotal;
    dataStream >> _coordinateSystem;
    dataStream >> _numberOfGridAxes;
    dataStream >> _constantGrid;
    _environmentType.unmarshal(dataStream);
    _orientation.unmarshal(dataStream);
    dataStream >> _sampleTime;
    dataStream >> _totalValues;
    dataStream >> _vectorDimension;
    dataStream >> _padding1;
    dataStream >> _padding2;

     _gridDataList.clear();
     for(size_t idx = 0; idx < _numberOfGridAxes; idx++)
     {
        GridAxisRecord x;
        x.unmarshal(dataStream);
        _gridDataList.push_back(x);
     }
}


bool GriddedDataPdu::operator ==(const GriddedDataPdu& rhs) const
 {
     bool ivarsEqual = true;

     ivarsEqual = SyntheticEnvironmentFamilyPdu::operator==(rhs);

     if( ! (_environmentalSimulationApplicationID == rhs._environmentalSimulationApplicationID) ) ivarsEqual = false;
     if( ! (_fieldNumber == rhs._fieldNumber) ) ivarsEqual = false;
     if( ! (_pduNumber == rhs._pduNumber) ) ivarsEqual = false;
     if( ! (_pduTotal == rhs._pduTotal) ) ivarsEqual = false;
     if( ! (_coordinateSystem == rhs._coordinateSystem) ) ivarsEqual = false;
     if( ! (_constantGrid == rhs._constantGrid) ) ivarsEqual = false;
     if( ! (_environmentType == rhs._environmentType) ) ivarsEqual = false;
     if( ! (_orientation == rhs._orientation) ) ivarsEqual = false;
     if( ! (_sampleTime == rhs._sampleTime) ) ivarsEqual = false;
     if( ! (_totalValues == rhs._totalValues) ) ivarsEqual = false;
     if( ! (_vectorDimension == rhs._vectorDimension) ) ivarsEqual = false;
     if( ! (_padding1 == rhs._padding1) ) ivarsEqual = false;
     if( ! (_padding2 == rhs._padding2) ) ivarsEqual = false;

     for(size_t idx = 0; idx < _gridDataList.size(); idx++)
     {
        if( ! ( _gridDataList[idx] == rhs._gridDataList[idx]) ) ivarsEqual = false;
     }


    return ivarsEqual;
 }

int GriddedDataPdu::getMarshalledSize() const
{
   int marshalSize = 0;

   marshalSize = SyntheticEnvironmentFamilyPdu::getMarshalledSize();
   marshalSize = marshalSize + _environmentalSimulationApplicationID.getMarshalledSize();  // _environmentalSimulationApplicationID
   marshalSize = marshalSize + 2;  // _fieldNumber
   marshalSize = marshalSize + 2;  // _pduNumber
   marshalSize = marshalSize + 2;  // _pduTotal
   marshalSize = marshalSize + 2;  // _coordinateSystem
   marshalSize = marshalSize + 1;  // _numberOfGridAxes
   marshalSize = marshalSize + 1;  // _constantGrid
   marshalSize = marshalSize + _environmentType.getMarshalledSize();  // _environmentType
   marshalSize = marshalSize + _orientation.getMarshalledSize();  // _orientation
   marshalSize = marshalSize + 8;  // _sampleTime
   marshalSize = marshalSize + 4;  // _totalValues
   marshalSize = marshalSize + 1;  // _vectorDimension
   marshalSize = marshalSize + 2;  // _padding1
   marshalSize = marshalSize + 1;  // _padding2

   for(int idx=0; idx < _gridDataList.size(); idx++)
   {
        GridAxisRecord listElement = _gridDataList[idx];
        marshalSize = marshalSize + listElement.getMarshalledSize();
    }

    return marshalSize;
}

